org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestManager.xtend

```diff
@@ -7,111 +7,134 @@
  *******************************************************************************/
 package org.eclipse.xtext.ide.server.concurrent
 
+import com.google.common.util.concurrent.ThreadFactoryBuilder
 import com.google.inject.Inject
 import java.util.concurrent.CompletableFuture
 import java.util.concurrent.CompletionException
-import java.util.concurrent.CopyOnWriteArrayList
 import java.util.concurrent.ExecutorService
-import java.util.concurrent.locks.Lock
-import java.util.concurrent.locks.ReentrantReadWriteLock
+import java.util.concurrent.Executors
+import org.eclipse.xtend.lib.annotations.FinalFieldsConstructor
 import org.eclipse.xtext.service.OperationCanceledManager
 import org.eclipse.xtext.util.CancelIndicator
 
 /**
- * <p>
- * The request manager run read/write requests.
- * </p>
- * <p>
- * <b>Execution:</b>
- * <ul>
- *  <li>requests are processed in an insertion order;</li>
- * 	<li>write requests are executed exclusive;</li>
- *  <li>read requests are executed in parallel.</li>
- * </ul>
- * </p>
  * 
  * @author kosyakov - Initial contribution and API
  * @since 2.11
  */
 class RequestManager {
 
-	private final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
-	private final Lock r = rwl.readLock();
-	private final Lock w = rwl.writeLock();
-
 	@Inject ExecutorService parallel
 
 	@Inject
 	OperationCanceledManager operationCanceledManager
 
-	val toCancel = new CopyOnWriteArrayList<Cancellable>
+	val queue = Executors.newSingleThreadExecutor(
+		new ThreadFactoryBuilder().setDaemon(true).setNameFormat("RequestManager-Queue-%d").build
+	)
+	var requests = <AbstractRequest<?>>newArrayList
 
 	def void shutdown() {
+		queue.shutdown()
 		parallel.shutdown()
 		cancel()
 	}
 
-	def <V> CompletableFuture<V> runRead((CancelIndicator)=>V request) {
-		val result = new CompletableFuture<V>()
-		writePending.thenRun [			
-			parallel.submit[
-				val cancelIndicator = new RequestCancelIndicator(result)
-				try {
-					r.lock
-					toCancel += cancelIndicator
-					cancelIndicator.checkCanceled
-					result.complete(request.apply(cancelIndicator))
-				} catch (Throwable e) {
-					result.completeExceptionally(e)
-				} finally {
-					toCancel -= cancelIndicator
-					r.unlock()
-				}
-			]
-		]
-		return result
+	def <V> CompletableFuture<V> runRead((CancelIndicator)=>V cancellable) {
+		return submit(
+			new ReadRequest(cancellable, parallel)
+		)
+	}
+
+	def <U, V> CompletableFuture<V> runWrite(()=>U nonCancellable, (CancelIndicator, U)=>V cancellable) {
+		val cancelFuture = cancel()
+		return submit(
+			new WriteRequest(nonCancellable, cancellable, cancelFuture)
+		)
+	}
+
+	protected def <V> CompletableFuture<V> submit(AbstractRequest<V> request) {
+		requests += request
```

Comment:
Can this method run concurrently? If yes, there will be concurrent add's to an ArrayList which will leave the structure in a broken state.

Reply:
no, it can not be, requests are queued sequentially

Reply:
Ok, I guess I misinterpreted the package `concurrent` :)

Reply:
it gets concurrent later in execution of read requests :P